File: simulator/src/VehicleSimulator.cpp
#include "VehicleSimulator.h"
#include <cmath>
#include <iostream>

namespace AetherDrive {

VehicleSimulator::VehicleSimulator() {
    m_telemetry.speed = 0.0;
    m_telemetry.rpm = 800;
    m_telemetry.battery = 85;
    m_telemetry.range = static_cast<int>(m_batteryFloat * 4.1);
    m_telemetry.gear = "P";
    m_telemetry.doorLocked = true;
    m_telemetry.driveMode = "NORMAL";
    m_telemetry.speedLimit = 100.0;
}

VehicleSimulator::~VehicleSimulator() {
    stop();
}

void VehicleSimulator::setCallback(TelemetryCallback cb) {
    m_callback = std::move(cb);
}

void VehicleSimulator::start() {
    if (m_running.load()) return;
    m_running.store(true);
    m_workerThread = std::thread(&VehicleSimulator::runLoop, this);
}

void VehicleSimulator::stop() {
    if (!m_running.load()) return;
    m_running.store(false);
    if (m_workerThread.joinable()) {
        m_workerThread.join();
    }
}

bool VehicleSimulator::isRunning() const {
    return m_running.load();
}

void VehicleSimulator::setTargetSpeed(double targetSpeedKmH) {
    m_targetSpeed = targetSpeedKmH;
}

void VehicleSimulator::setGear(const std::string& gear) {
    m_telemetry.gear = gear;
}

void VehicleSimulator::triggerLaneDeparture(bool active) {
    m_laneAlertActive = active;
}

void VehicleSimulator::updatePhysics(double dt) {
    m_timeElapsed += dt;

    // Natural highway driving speed profile simulation (0 -> 45 -> 75 -> 105 -> 65 -> 80)
    // Sine harmonic wave creates realistic traffic variations
    double speedOscillation = 18.0 * std::sin(m_timeElapsed * 0.15) + 8.0 * std::cos(m_timeElapsed * 0.35);
    double desiredSpeed = std::max(0.0, m_targetSpeed + speedOscillation);

    // Initial warm-up gear shifting
    if (m_timeElapsed < 2.0) {
        m_telemetry.gear = "P";
        desiredSpeed = 0.0;
    } else if (m_telemetry.gear == "P") {
        m_telemetry.gear = "D";
    }

    // Smooth speed change
    if (m_telemetry.speed < desiredSpeed) {
        m_telemetry.speed += m_accelerationRate * dt * 4.0;
        if (m_telemetry.speed > desiredSpeed) m_telemetry.speed = desiredSpeed;
    } else if (m_telemetry.speed > desiredSpeed) {
        m_telemetry.speed -= m_accelerationRate * dt * 3.0; // deceleration
        if (m_telemetry.speed < desiredSpeed) m_telemetry.speed = desiredSpeed;
    }

    // RPM calculations: base motor/engine rpm directly scaled with speed
    if (m_telemetry.speed < 1.0) {
        m_telemetry.rpm = 800; // Idle
    } else {
        // Gear ratio simulation
        double baseRpm = 1000.0 + (m_telemetry.speed * 42.0);
        // Add subtle engine flutter
        double jitter = 35.0 * std::sin(m_timeElapsed * 1.5);
        m_telemetry.rpm = static_cast<int>(baseRpm + jitter);
    }

    // Battery & Regenerative Braking
    if (m_telemetry.speed > desiredSpeed) {
        // Regenerative braking: slight recharge
        m_batteryFloat += 0.002 * dt;
    } else {
        // Power consumption based on speed squared
        double powerDrain = (0.012 + (m_telemetry.speed / 140.0) * 0.02) * dt;
        m_batteryFloat -= powerDrain;
    }
    if (m_batteryFloat < 5.0) m_batteryFloat = 98.0; // reset for continuous demo

    m_telemetry.battery = static_cast<int>(std::round(m_batteryFloat));
    m_telemetry.range = static_cast<int>(m_batteryFloat * 4.1);

    // Auto door locking above 15 km/h
    m_telemetry.doorLocked = (m_telemetry.speed > 15.0);

    // Dynamic alerts
    m_telemetry.laneDepartureWarning = m_laneAlertActive || (std::fmod(m_timeElapsed, 45.0) > 40.0);
    if (m_telemetry.speed > m_telemetry.speedLimit + 10.0) {
        m_telemetry.alertMessage = "Overspeed Warning: Exceeding Road Limit";
    } else if (m_telemetry.battery < 20) {
        m_telemetry.alertMessage = "Low Battery: Find Nearest Charger";
    } else if (m_telemetry.laneDepartureWarning) {
        m_telemetry.alertMessage = "Lane Departure: Keep Vehicle Centered";
    } else {
        m_telemetry.alertMessage = "";
    }

    m_telemetry.timestamp = VehicleTelemetry::currentIsoTimestamp();
}

void VehicleSimulator::runLoop() {
    const double dt = 0.1; // 100ms update rate = 10 Hz
    while (m_running.load()) {
        updatePhysics(dt);

        if (m_callback) {
            m_callback(m_telemetry);
        }

        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }
}

} // namespace AetherDrive

File: simulator/vehicle_simulator.py
#!/usr/bin/env python3
"""
AetherDrive - SDV Vehicle Telemetry Simulator (Python + MQTT)
Publishes realistic vehicle telemetry data via MQTT to topic 'aetherdrive/vehicle/data'.
"""

import argparse
import datetime
import json
import math
import sys
import time
import paho.mqtt.client as mqtt

DEFAULT_BROKER = "broker.emqx.io"
DEFAULT_PORT = 1883
TOPIC_VEHICLE_DATA = "aetherdrive/vehicle/data"
TOPIC_VEHICLE_ALERT = "aetherdrive/vehicle/alert"

class VehicleSimulator:
    def __init__(self, broker_host=DEFAULT_BROKER, broker_port=DEFAULT_PORT):
        self.broker_host = broker_host
        self.broker_port = broker_port

        # Vehicle state
        self.speed = 0.0          # km/h
        self.target_speed = 75.0  # cruising target
        self.rpm = 800            # RPM
        self.battery = 88.0       # %
        self.range_km = 360       # km
        self.gear = "P"           # P, R, N, D
        self.door_locked = True
        self.speed_limit = 100
        self.drive_mode = "NORMAL"
        self.lane_departure = False
        self.alert_msg = ""
        
        self.time_elapsed = 0.0
        self.client = None
        self.connected = False

    def on_connect(self, client, userdata, flags, rc, properties=None):
        if rc == 0:
            self.connected = True
            print(f"[SIMULATOR] Connected successfully to MQTT broker {self.broker_host}:{self.broker_port}")
        else:
            print(f"[SIMULATOR] MQTT Connection returned result code {rc}")

    def on_disconnect(self, client, userdata, disconnect_flags_or_rc, reason_code=None, properties=None):
        self.connected = False
        print("[SIMULATOR] Disconnected from MQTT broker. Retrying in background...")

    def connect_mqtt(self):
        try:
            if self.client:
                try:
                    self.client.loop_stop()
                    self.client.disconnect()
                except Exception:
                    pass

            try:
                self.client = mqtt.Client(mqtt.CallbackAPIVersion.VERSION2, "AetherDrive-Simulator")
            except AttributeError:
                self.client = mqtt.Client("AetherDrive-Simulator")

            self.client.reconnect_delay_set(min_delay=1, max_delay=30)
            self.client.on_connect = self.on_connect
            self.client.on_disconnect = self.on_disconnect
            
            print(f"[SIMULATOR] Connecting to MQTT broker {self.broker_host}:{self.broker_port}...")
            self.client.connect_async(self.broker_host, self.broker_port, 60)
            self.client.loop_start()
        except Exception as e:
            print(f"[SIMULATOR] Warning: Could not connect to MQTT broker ({e}). Will retry automatically.")

    def update_physics(self, dt=0.1):
        self.time_elapsed += dt

        # Realistic dynamic driving cycle:
        # 0 - 3s: Parked
        # 3s+: Drive mode with harmonic acceleration/braking curves
        if self.time_elapsed < 3.0:
            self.gear = "P"
            desired_speed = 0.0
        else:
            self.gear = "D"
            # Sine wave speed fluctuation simulating city + highway transitions
            fluctuation = 22.0 * math.sin(self.time_elapsed * 0.12) + 10.0 * math.cos(self.time_elapsed * 0.3)
            desired_speed = max(0.0, self.target_speed + fluctuation)

        # Acceleration and deceleration curves
        if self.speed < desired_speed:
            self.speed = min(desired_speed, self.speed + (3.8 * dt * 3.5))
        elif self.speed > desired_speed:
            self.speed = max(desired_speed, self.speed - (2.5 * dt * 3.5))

        # Motor / Engine RPM calculation
        if self.speed < 1.0:
            self.rpm = 800
        else:
            base_rpm = 950 + (self.speed * 42.0)
            flutter = 30.0 * math.sin(self.time_elapsed * 2.0)
            self.rpm = int(base_rpm + flutter)

        # Battery drainage & Regenerative braking
        if self.speed > desired_speed:
            # Regenerative braking slightly recovers charge
            self.battery = min(100.0, self.battery + (0.003 * dt))
        else:
            # Power consumption
            self.battery = max(2.0, self.battery - ((0.01 + (self.speed / 120.0) * 0.018) * dt))

        self.range_km = int(self.battery * 4.1)
        self.door_locked = (self.speed > 15.0)

        # Periodic ADAS lane departure alert simulation every ~40 seconds for 4 seconds
        cycle = self.time_elapsed % 45.0
        self.lane_departure = (40.0 <= cycle <= 44.0)

        # Alerts
        if self.speed > self.speed_limit + 12:
            self.alert_msg = "OVERSPEED: Reduce speed below limit"
        elif self.battery < 20.0:
            self.alert_msg = "BATTERY LOW: Plan charging stop"
        elif self.lane_departure:
            self.alert_msg = "LANE DEPARTURE: Keep centered"
        else:
            self.alert_msg = ""

    def generate_packet(self):
        iso_timestamp = datetime.datetime.now(datetime.timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
        return {
            "speed": int(round(self.speed)),
            "rpm": int(self.rpm),
            "battery": int(round(self.battery)),
            "range": int(self.range_km),
            "gear": self.gear,
            "door_locked": self.door_locked,
            "speed_limit": self.speed_limit,
            "drive_mode": self.drive_mode,
            "lane_departure": self.lane_departure,
            "alert": self.alert_msg,
            "timestamp": iso_timestamp
        }

    def run(self, interval=0.1):
        self.connect_mqtt()
        print("\n[SIMULATOR] Vehicle Telemetry Generator started (10 Hz).")
        print(f"[SIMULATOR] Publishing to topic: '{TOPIC_VEHICLE_DATA}'")
        print("[SIMULATOR] Press Ctrl+C to stop.\n")

        counter = 0
        try:
            while True:
                self.update_physics(interval)
                packet = self.generate_packet()
                payload = json.dumps(packet)

                if self.client and self.connected:
                    self.client.publish(TOPIC_VEHICLE_DATA, payload, qos=0)
                    if packet["alert"]:
                        self.client.publish(TOPIC_VEHICLE_ALERT, json.dumps({
                            "type": "ALERT",
                            "message": packet["alert"],
                            "timestamp": packet["timestamp"]
                        }), qos=1)

                counter += 1
                if not self.connected and counter % 50 == 0:
                    self.connect_mqtt()

                if counter % 10 == 0:  # Log to console once per second
                    status = "CONNECTED" if self.connected else "BROADCASTING"
                    alert_tag = f" | [!] {packet['alert']}" if packet["alert"] else ""
                    print(f"[{status}] Speed: {packet['speed']:3d} km/h | RPM: {packet['rpm']:4d} | "
                          f"Battery: {packet['battery']:2d}% | Range: {packet['range']:3d} km | "
                          f"Gear: {packet['gear']}{alert_tag}", flush=True)

                time.sleep(interval)
        except KeyboardInterrupt:
            print("\n[SIMULATOR] Stopping vehicle simulator...", flush=True)
        finally:
            if self.client:
                self.client.loop_stop()
                self.client.disconnect()
            print("[SIMULATOR] Done.", flush=True)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="AetherDrive SDV Vehicle Telemetry Simulator")
    parser.add_argument("--broker", default=DEFAULT_BROKER, help="MQTT Broker host (default: broker.emqx.io)")
    parser.add_argument("--port", type=int, default=DEFAULT_PORT, help="MQTT Broker port (default: 1883)")
    args = parser.parse_args()

    sim = VehicleSimulator(args.broker, args.port)
    sim.run()
